File: FD/C#/utils/ImageLoader.cs
using System;
using System.Runtime.InteropServices;
using System.Drawing;
using System.Drawing.Imaging;

namespace arcsoft {
    namespace utils {
        public class ImageLoader {
            public static BufferInfo getI420FromFile(String filePath) {
                byte[] yuv = null;
                int w = 0;
                int h = 0;
                {
                    Bitmap img = new Bitmap(filePath);
                    if (((img.Width & 0x1) != 0) || ((img.Height & 0x1) != 0)) {
                        img = img.Clone(new RectangleF(0, 0, img.Width & 0xFFFFFFFE, img.Height & 0xFFFFFFFE), PixelFormat.Format32bppRgb);
                    }
                    w = img.Width;
                    h = img.Height;


                    int[] bgra = new int[img.Width * img.Height];
                    Rectangle rect = new Rectangle(0, 0, img.Width, img.Height);
                    BitmapData bmpData = img.LockBits(rect, System.Drawing.Imaging.ImageLockMode.ReadOnly, PixelFormat.Format32bppRgb);
                    Marshal.Copy(bmpData.Scan0, bgra, 0, img.Width * img.Height );
                    img.UnlockBits(bmpData);

                    yuv = BGRA2I420(bgra, w, h);
                }
                return new BufferInfo(w, h, yuv);
            }

            public static byte[] BGRA2I420(int[] bgra, int width, int height) {

                byte[] yuv = new byte[width * height * 3 / 2];
                int u_offset = width * height;
                int y_offset = width * height * 5 / 4;

                for (int i = 0; i < height; i++) {
                    for (int j = 0; j < width; j++) {
                        int rgb = bgra[i * width + j] & 0x00FFFFFF;
                        int b = rgb & 0xFF;
                        int g = (rgb >> 8) & 0xFF;
                        int r = (rgb >> 16) & 0xFF;

                        int y = ((66 * r + 129 * g + 25 * b + 128) >> 8) + 16;
                        int u = ((-38 * r - 74 * g + 112 * b + 128) >> 8) + 128;
                        int v = ((112 * r - 94 * g - 18 * b + 128) >> 8) + 128;

                        y = y < 0 ? 0 : (y > 255 ? 255 : y);
                        u = u < 0 ? 0 : (u > 255 ? 255 : u);
                        v = v < 0 ? 0 : (v > 255 ? 255 : v);

                        yuv[i * width + j] = (byte)y;
                        yuv[u_offset + (i >> 1) * (width >> 1) + (j >> 1)] = (byte)u;
                        yuv[y_offset + (i >> 1) * (width >> 1) + (j >> 1)] = (byte)v;
                    }
                }
                return yuv;
            }
        }
    }
}

File: FR/python/arcsoft/utils/ImageLoader.py
#-*- encoding=utf-8 -*-
from PIL import Image
from . import BufferInfo

def BGRA2I420(img, width, height):
    yuv = bytearray(width * height * 3 // 2)
    u_offset = width * height
    y_offset = width * height * 5 // 4

    for i in range(0, height):
        for j in range(0, width):
            rgb = img.getpixel((j,i))
            r = rgb[0]
            g = rgb[1]
            b = rgb[2]

            y = ((66 * r + 129 * g + 25 * b + 128) >> 8) + 16
            u = ((-38 * r - 74 * g + 112 * b + 128) >> 8) + 128
            v = ((112 * r - 94 * g - 18 * b + 128) >> 8) + 128

            y = 0 if y < 0 else (255 if y > 255 else (y & 0xFF))
            u = 0 if u < 0 else (255 if u > 255 else (u & 0xFF))
            v = 0 if v < 0 else (255 if v > 255 else (v & 0xFF))

            yuv[i * width + j] = y
            yuv[u_offset + (i >> 1) * (width >> 1) + (j >> 1)] = u
            yuv[y_offset + (i >> 1) * (width >> 1) + (j >> 1)] = v

    return bytes(yuv)


def getI420FromFile(filePath):
    img = Image.open(filePath).convert('RGB')
    w = img.width& 0xFFFFFFFE
    h = img.height& 0xFFFFFFFE
    yuv = BGRA2I420(img, w, h)

    return BufferInfo(w, h, yuv)
